File: hoffman-agents-python/src/conscious_agent/core/token_inventor.py
from __future__ import annotations

import random

CONSONANTS = "bdfghjklmnprstvwxyz"
VOWELS = "aeiou"


def invent_token(rng: random.Random | None = None) -> str:
    if rng is None:
        rng = random
    c1 = rng.choice(CONSONANTS)
    v = rng.choice(VOWELS)
    c2 = rng.choice(CONSONANTS)
    return f"{c1}{v}{c2}"


def is_invented_token(token: str) -> bool:
    if token in ("I", "notice", "familiar", "different", "wait"):
        return False
    if token.startswith("p:"):
        return False
    if len(token) != 3:
        return False
    return (token[0] in CONSONANTS and token[1] in VOWELS and token[2] in CONSONANTS)

File: hoffman-agents-react-native/src/core/token-inventor.js
const crypto = require('../crypto');

const CONSONANTS = 'bdfghjklmnprstvwxyz';
const VOWELS = 'aeiou';
const CORE_TOKENS = new Set(['I', 'notice', 'familiar', 'different', 'wait']);

function inventToken() {
  const c1 = CONSONANTS[Math.floor(Math.random() * CONSONANTS.length)];
  const v = VOWELS[Math.floor(Math.random() * VOWELS.length)];
  const c2 = CONSONANTS[Math.floor(Math.random() * CONSONANTS.length)];
  return `${c1}${v}${c2}`;
}

function isInventedToken(token) {
  if (CORE_TOKENS.has(token)) return false;
  if (token.startsWith('p:')) return false;
  if (token.length !== 3) return false;
  return CONSONANTS.includes(token[0]) && VOWELS.includes(token[1]) && CONSONANTS.includes(token[2]);
}

module.exports = { inventToken, isInventedToken };
